sdk/lib/crt/except/amd64/ehandler.c

```diff
@@ -1,19 +1,143 @@
+
+typedef struct _C_SCOPE_TABLE
+{
+    ULONG NumEntries;
+    C_SCOPE_TABLE_ENTRY Entry[1];
+} C_SCOPE_TABLE, *PC_SCOPE_TABLE;
 
 
 _CRTIMP
 EXCEPTION_DISPOSITION
 __cdecl
 __C_specific_handler(
-    struct _EXCEPTION_RECORD *_ExceptionRecord,
-    void *_EstablisherFrame,
-    struct _CONTEXT *_ContextRecord,
-    struct _DISPATCHER_CONTEXT *_DispatcherContext)
+    struct _EXCEPTION_RECORD *ExceptionRecord,
+    void *EstablisherFrame,
+    struct _CONTEXT *ContextRecord,
+    struct _DISPATCHER_CONTEXT *DispatcherContext)
 {
-    UNIMPLEMENTED;
-    __debugbreak();
-    return 0;
+    PSCOPE_TABLE ScopeTable;
+    ULONG i, BeginAddress, EndAddress, Handler;
+    ULONG64 ImageBase, JumpTarget, IpOffset, TargetIpOffset;
+    EXCEPTION_POINTERS ExceptionPointers;
+    PTERMINATION_HANDLER TerminationHandler;
+    PEXCEPTION_FILTER ExceptionFilter;
+    LONG FilterResult;
+
+    /* Set up the EXCEPTION_POINTERS */
+    ExceptionPointers.ExceptionRecord = ExceptionRecord;
+    ExceptionPointers.ContextRecord = ContextRecord;
+
+    /* Get the image base */
+    ImageBase = (ULONG64)DispatcherContext->ImageBase;
+
+    /* Get the image base relative instruction pointers */
+    IpOffset = DispatcherContext->ControlPc - ImageBase;
+    TargetIpOffset = DispatcherContext->TargetIp - ImageBase;
+
+    /* Get the scope table and current index */
+    ScopeTable = (PSCOPE_TABLE)DispatcherContext->HandlerData;
+
+    /* Loop while we have scope table entries */
+    while (DispatcherContext->ScopeIndex < ScopeTable->Count)
+    {
+        /* Use i as index and update the dispatcher context */
+        i = DispatcherContext->ScopeIndex++;
+
+        /* Get the start and end of the scrope */
+        BeginAddress = ScopeTable->ScopeRecord[i].BeginAddress;
+        EndAddress = ScopeTable->ScopeRecord[i].EndAddress;
+
+        /* Skip this scope if we are not within the bounds */
+        if ((IpOffset < BeginAddress) || (IpOffset >= EndAddress))
+        {
+            continue;
+        }
+
+        /* Check if this is an unwind */
+        if (ExceptionRecord->ExceptionFlags & EXCEPTION_UNWIND)
+        {
+            /* Check if this is a target unwind */
+            if (ExceptionRecord->ExceptionFlags & EXCEPTION_TARGET_UNWIND)
+            {
+                /* Check if the target is within the scope itself */
+                if ((TargetIpOffset >= BeginAddress) &&
+                    (TargetIpOffset <  EndAddress))
+                {
+                    return ExceptionContinueSearch;
+                }
+            }
+
+            /* Check if this is a termination handler / finally function */
+            if (ScopeTable->ScopeRecord[i].JumpTarget == 0)
+            {
+                /* Call the handler */
+                Handler = ScopeTable->ScopeRecord[i].HandlerAddress;
+                TerminationHandler = (PTERMINATION_HANDLER)(ImageBase + Handler);
+                TerminationHandler(TRUE, EstablisherFrame);
```

Comment:
Are we sure that `TerminationHandler` will never ever be NULL? (could be NULL if the exception record gets corrupted)

Reply:
It could be 0xDEADBEEF, if your CPU or memory are corrupted. I'll add a check and send a message out on the keyboard LEDs in Morse code...

Reply:
I was just thinking about a corrupted executable file.

Reply:
Then you'll crash in your crash handler and this will be hard to debug. Life of someone writing a compiler or a malware, I guess...